src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,466 @@
+                if (dominates(n.getBlock().ptr, node.getBlock().ptr)) {
+                    nodeValues.put(node, n);
+                } else {
+                    nodeValues.put(n, node);
+                    cmpCache.put(cmpPreds, node);
+                }
+            } else {
+                cmpCache.put(cmpPreds, node);
+            }
+        }
+
+        @Override
+        public void visit(Cond node) {
+            // do nothing
+        }
+
+        @Override
+        public void visit(Proj proj) {
+            ProjPreds projPreds = new ProjPreds(proj.getPred(), proj.getNum(), proj.getMode());
+            if (projCache.containsKey(projPreds)) {
+                Node n = projCache.get(projPreds);
+                if (dominates(n.getBlock().ptr, proj.getBlock().ptr)) {
+                    nodeValues.put(proj, n);
+                } else {
+                    nodeValues.put(n, proj);
+                    projCache.put(projPreds, proj);
+                }
+            } else {
+                projCache.put(projPreds, proj);
+            }
+        }
+
+        @Override
+        public void visit(Address addr) {
+            Entity ent = addr.getEntity();
+            if (addressCache.containsKey(ent)) {
+                Node newAddr = addressCache.get(ent);
+                Graph.exchange(addr, newAddr);
+            } else {
+                addressCache.put(ent, addr);
+            }
+        }
+
+        @Override
+        public void visit(Call call) {
+            Attributes a = s.getAttributeAnalysis().getAttributes(Util.getCallee(call));
+            if (a.isPure()) {
+                visitPreds(call, Util.iterableToArray(call.getPreds()), binding_irnode.ir_opcode.iro_Call);
+            }
+        }
+
+        @Override
+        public void visit(Const node) {
+            TargetValue tarVal = node.getTarval();
+            if (constCache.containsKey(tarVal)) {
+                Node newConst = constCache.get(tarVal);
+                Graph.exchange(node, newConst);
+            } else {
+                constCache.put(tarVal, node);
+            }
+        }
+
+        @Override
+        public void visit(Conv node) {
+            visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Conv);
+        }
+
+        @Override
+        public void visit(Div node) {
+            //   visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Div);
+        }
+
+        @Override
+        public void visit(Eor node) {
+            visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Eor);
+        }
+
+        @Override
+        public void visit(Minus node) {
+            visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Minus);
+        }
+
+        @Override
+        public void visit(Mod node) {
+            //   visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Mod);
```

Comment:
`Mod` and `Div` are still commented out.

Reply:
Done